Fix Vulkan CPU/GPU synchronization

darksylinc opened the pull request:
The original code had two bugs:

1. It assumed Swapchain count can be used for proper CPU/GPU synchronization (it's not)
2. It used a double buffer scheme (FRAME_LAG = 2) but the vkWaitForFences/vkResetFences calls were in the wrong order and in the wrong place
3. The swapchain bug forced a quad-buffer scheme with roughly 1 or 2 frames of wait; which hid all the issues

This commit gets rid of `RenderingDeviceVulkan::frame_count` & `RenderingDeviceVulkan::frame` which should've never existed and makes everything depend on `VulkanContext::frame_index` &
`VulkanContext::FRAME_LAG`

See https://github.com/godotengine/godot/issues/80550 for details

Fix #80550
Likely fixes https://github.com/godotengine/godot/issues/81670.

## Specific things that need testing:

Aside from overall stability and ensure there are no regressions

- [ ] See if there is an improvement for DEVICE_LOST users. #71929
- [ ]  See if users who reported higher latency/lag vs OpenGL renderer report improvements #71795
- [x] Test if performance changes (for better or worse)
- [x] Test FRAME_LAG = 3. The buggy behavior had the latency disadvantages of triple buffering with the performance disadvantages of double buffering. Perhaps we can get a speed boost with the same latency as before by using FRAME_LAG = 3

**Note:** This PR may uncover [bugs that were previously hidden](https://github.com/godotengine/godot/issues/80570) or that were rare to reproduce. This may give the illusion that the code has become more unstable when in reality it's just uncovering hidden flaws.

BastiaanOlij commented:
I need to test this sometime during the week with VR. I've had a lot of timing issues because our frame timing seemed to clash with OpenXRs `xrWaitFrame` logic, or at least thats always been my conclusion, but possibly this is the real issue.

darksylinc commented:
OK I have evidence that lag has been reduced.

I couldn't use OBS because it introduced atrocious lag when enabled. I had to use an external camera.

I've ran multiple non-scientific tests and it felt better but I was heavily biased (I *want* to claim my version does something); so I had to 

# Methodology

1. Xubuntu 20.04 XFCE. No Compositor (Window Manager Tweaks -> Compositor -> Enable display compositing must be off)
2. X11 - Vulkan (Forward+) - dedicated AMD Radeon RX 6800 XT - AMD Ryzen 9 5900X 12-Core Processor (24 Threads)
3. Using RADV driver. Mesa 23.0.4 - kisak-mesa PPA
4. TearFree enabled (xrandr --output DisplayPort-1 -set TearFree on)
5. LG FLATRON E2342 1920x1080@60hz
6. Open near-empty project ([PopupCrash](https://github.com/godotengine/godot/files/8459740/PopupCrash.zip)).
7. Camera used for recording: Xiaomi POCO F2 set to 1080p@60 FPS
8. VSync on
9. Update Continuously is off
10. Mouse movement is emulated with the following bash script (use `sudo apt install xautomation` if not installed):
```bash
while true
do
	xte 'mousemove 100 100'
	xte 'mousemove 100 125'
	xte 'mousemove 100 150'
	xte 'mousemove 100 175'
	xte 'mousemove 100 200'
	xte 'mousemove 100 225'
	xte 'mousemove 100 250'
	xte 'mousemove 100 275'
	xte 'mousemove 100 300'
	xte 'mousemove 100 325'
	xte 'mousemove 100 350'
	xte 'mousemove 100 325'
	xte 'mousemove 100 300'
	xte 'mousemove 100 275'
	xte 'mousemove 100 250'
	xte 'mousemove 100 225'
	xte 'mousemove 100 200'
	xte 'mousemove 100 175'
	xte 'mousemove 100 150'
	xte 'mousemove 100 125'
	xte 'mousemove 100 100'
done
```

These are the results (I've had to re-encode the videos because they were too big for Github):

**Before PR**
https://github.com/godotengine/godot/assets/3395130/63a681df-98e6-4836-aa24-cf0e43977b58

**After PR**
https://github.com/godotengine/godot/assets/3395130/d5b40e06-23f5-4742-8392-5cd9cb674bd8


**We can notice:**

1. Before PR the highlight lags behind a lot
2. Before PR the hightlight is almost always 9 highlights behind (when the cursor is in "New Inherited Scene..." the highlight is in "Quick Open Scene...")
3. After PR lags behind a lot less
4. Even though in a few occasions it manages to be 9 highlights behind, it happens twice in the video, while most often is only at most 7 highlights behind; while most of the time is 3 highlights behind.

Overall I believe the data is consistent with latency being reduced by 1 frame (16 ms). I encourage others to do their own research and validate the result.

Unfortunately I only run the test once (I'm a bit of a hurry, tomorrow it's election day).

IMHO it feels noticeable snappier too. Of course it will never reach the level of snappiness you can get with VSync Off, however theoretically the latency with VSync off should be reduced as well.

clayjohn commented on drivers/vulkan/vulkan_context.cpp:
```diff
@@ -2509,6 +2506,11 @@ Error VulkanContext::swap_buffers() {
 
+	// We must wait for the tail frame_index to finish rendering in the GPU, otherwise its resources
+	// (GPU memory address ranges + API handles) may still be in use.
+	// Ideally we'd dellay calling this as much as possible; but it's hard to guarantee.
```
```suggestion
	// Ideally we'd delay calling this as long as possible; but it's hard to guarantee.
```

clayjohn commented on drivers/vulkan/rendering_device_vulkan.cpp:
```diff
@@ -8933,11 +8968,8 @@ void RenderingDeviceVulkan::initialize(VulkanContext *p_context, bool p_local_de
 	context = p_context;
 	device = p_context->get_device();
 	if (p_local_device) {
-		frame_count = 1;
```
The implications of removing this aren't totally clear to me. The logic with p_local_device is slightly different as it isn't connected to a window in that case. I'll have to think about this more.

darksylinc replied:
Hi!

There are two independent things going on:

1. What frame_count was supposed to do. I've been meaning to ask you that. (and it sounds like this should be user-provided rather than assuming a value).
      - Right now `FRAME_LAG` does the job of the max frames to be ahead. It is hardcoded but I will be modifying this to be user-tweakable
2. `frame_count` was completely broken as I described in #80550 .

clayjohn replied:
The issue I am referring to comes from the fact that the local device doesn't have a swapchain and it never calls swap_buffers. It is essentially one frame, and it has to call ``submit()`` then ``sync()`` in order to submit more operations. In those cases, frame_count should always be 1 as there will literally never be more than one "frame".

Obtaining a global frame count from the context fundamentally breaks thing for local devices. 

I'm not opposed to grabbing the frame count from context, but if we do that, then we need more checks if we are using a local device to ensure we aren't referencing resources that don't exist

clayjohn reviewed (commented):
Some initial code review. Overall, I think this change is fantastic. But we indeed need to get as much testing as possible

I tested locally on PopOS 22.04 with Intel integrated graphics, X11 WM. I didn't see any perceptible changes in performance/lag/stability. Although my experience was already quite good.

darksylinc commented:
# Downloads for testing

- [Windows](https://github.com/godotengine/godot/suites/15045770947/artifacts/859188376)
- [macOS](https://github.com/godotengine/godot/suites/15045770947/artifacts/859188373)
- [Linux](https://github.com/godotengine/godot/suites/15045770947/artifacts/859188370)

darksylinc commented:
I'm adding new functionality.

In `Window -> V-Sync -> Buffer Count` users can now set the FRAME_LAG (which was previously hardcoded) anywhere in range [1; 4].

Since it must be provided at very early initialization, I didn't add a gdscript interface for it. That can be added & figured out later by anyone else.

![Godot](https://github.com/godotengine/godot/assets/3395130/0d93b417-9104-4f31-bdc0-8ec38f5b1dee)

The plumbing for that is done now (you may have feedback for changes of course, though).

This setting should also address @clayjohn concerns; although they may be misplaced.

Even if Godot has no windows, `VulkanContext::swap_buffers()` still must be called. Whether frame_count is 1 or 2 is irrelevant to whether there are windows.
frame_count just means how much the CPU is allowed to get ahead of the GPU.

Calinou commented:
> Since it must be provided at very early initialization, I didn't add a gdscript interface for it. That can be added & figured out later by anyone else.

This may likely require something like https://github.com/godotengine/godot-proposalsI/issues/6423 to be changeable at run-time.

Calinou commented on doc/classes/ProjectSettings.xml:
```diff
@@ -830,6 +830,16 @@
 			[b]Note:[/b] This property is only read when the project starts. To change the V-Sync mode at runtime, call [method DisplayServer.window_set_vsync_mode] instead.
 		</member>
+		<member name="display/window/vsync/buffer_count" type="int" setter="" getter="" default="2">
+			Sets the number of buffers before stalling to wait for the GPU.
+			2 corresponds to double buffering and 3 to triple buffering.
+			Single buffering may give you the lowest lag but at the high cost of no parallelism between CPU &amp; GPU
+			Assuming you're able to reach 60fps all the time, then double buffering gives you the lowest lag.
+			Use Triple buffering if V-Sync is enabled and the GPU is struggling to keep 60 fps (or whatever the monitor refresh rate is).
+			Triple buffering is almost a must if you're using V-Sync MAILBOX mode.
+			[b]Note:[/b] This setting is not supported by all APIs.
+			[b]Note:[/b] This property is only read when the project starts.
```
Every line break creates a new paragraph, so I suggest using those sparingly.

```suggestion
			Sets the number of buffers before stalling to wait for the GPU. A higher values allows for higher framerates in some situations at the cost of higher visual latency. 2 corresponds to double buffering and 3 to triple buffering. Single buffering may give you the lowest latency but at the high cost of no parallelism between CPU &amp; GPU
			Assuming you're able to reach the monitor's refresh rate all the time, double buffering gives you the lowest latency while still having some level of parallelism.
			Use Triple buffering if V-Sync is enabled and the GPU is struggling to keep up with the monitor's refresh rate.
			Triple buffering is almost a must if you're using the Mailbox V-Sync mode.
			[b]Note:[/b] This setting is only supported with Vulkan-based rendering methods (Forward+ and Mobile), not Compatibility.
			[b]Note:[/b] This property is only read when the project starts. There is currently no way to change this value at run-time.
```

Also, are the terms "double buffering" and "triple buffering" correct here? I see those being used to refer to V-Sync behavior, with double buffering offering less latency but a "permanent" drop to 30 FPS (or 20 FPS, 15 FPS, …) whenever you can't sustain 60 FPS. I thought this was configured separately, see https://github.com/godotengine/godot-proposals/issues/4065.

Calinou commented on main/main.cpp:
```diff
@@ -1993,6 +1994,21 @@ Error Main::setup(const char *execpath, int argc, char *argv[], bool p_second_ph
+		//	   time while at other times it lacks enough time.
+		//
+		// Alternatively, it makes a lot of sense if set to VSYNC_ADAPTIVE or VSYNC_MAILBOX.
+		// When VSync off, triple buffering increases lag for no (or barely?) improvement and wastes VRAM.
+		// so we force double buffering on that case.
+		if (window_vsync_mode == DisplayServer::VSyncMode::VSYNC_DISABLED)
+			frame_buffering_count = MIN(uint8_t(2u), frame_buffering_count);
+	}
```
This behavior should be documented in the class reference – I had to read the source code to find out about it :slightly_smiling_face:

darksylinc replied:
> Single-buffering is apparently OK with V-Sync disabled if I'm reading the code right too. This makes sense in my experience with various games.

Single buffering is almost never OK as the CPU will never overlap with the GPU to run in parallel (it can be ok in very specific non-gaming scenarios where latency is very strict and the app has control on the HW it is being run on and the performance characteristics; otherwise it's almost always a bad idea).

Anyway I'll have to edit this PR to separate buffer count from swapchain count because the simulator results are really unforgiving tying 3 swapchains to buffer count = 2 adds a lot of lag.

Calinou replied:
> Single buffering is almost never OK as the CPU will never overlap with the GPU to run in parallel

By "OK", I meant "technically valid", rather than "advised" :slightly_smiling_face:

darksylinc replied:
Ah yes!

It is valid. That's why I forced the value to be in range [1; MAX_FRAME_LAG] (the value of MAX_FRAME_LAG is arbitrary but anything larger than that is probably useless).

darksylinc commented:
OK I've updated the PR with `buffer_count` and `swapchain_count` separately.

The documentation includes a link to the [V-Sync Simulator](https://darksylinc.github.io/vsync_simulator/) I wrote to explain how they work.

I think the PR is ready to merge. If everyone's ok with the review, I'll rebase all the commits to squash them so it can be merged.

AThousandShips commented:
Documentation must be sorted alphabetically

clayjohn commented on main/main.cpp:
```diff
@@ -1993,6 +1995,22 @@ Error Main::setup(const char *execpath, int argc, char *argv[], bool p_second_ph
 		}
 	}
+	{
+		frame_buffering_count = uint8_t(GLOBAL_DEF_BASIC("display/window/vsync/buffer_count", 2u));
+		swapchain_count = uint8_t(GLOBAL_DEF_BASIC("display/window/vsync/swapchain_count", 3u));
```
Since these can only be changed at startup, they should use ``GLOBAL_DEF_RST_BASIC`` which will prompt the user to restart if they adjust the setting. 

Personally, I think these should be ``GLOBAL_DEF_RST`` as they are settings for advanced users only. Removing BASIC means that they setting will only appear if the "advanced" toggle is toggled on.
```suggestion
		frame_buffering_count = uint8_t(GLOBAL_DEF_BASIC("display/window/vsync/buffer_count", 2u));
		swapchain_count = uint8_t(GLOBAL_DEF_BASIC("display/window/vsync/swapchain_count", 3u));
```

clayjohn commented on drivers/vulkan/vulkan_context.cpp:
```diff
@@ -2765,7 +2767,13 @@ void VulkanContext::set_vsync_mode(DisplayServer::WindowID p_window, DisplayServ
 }
 
-VulkanContext::VulkanContext() {
+VulkanContext::VulkanContext(const uint8_t p_frame_count, const uint8_t p_swapchain_count) {
```
Is there a reason to pass these values through the Context creation? You should be able to read the project settings here using ``GLOBAL_GET`` (or alternatively, you could fill these values in initialize as well)

If I understand correctly, the settings need to be created and validated in main.cpp. But I think we could rely on the project setting system to read them here instead of adding plumping throughout the displayserver

AThousandShips replied:
```suggestion
VulkanContext::VulkanContext(uint8_t p_frame_count, uint8_t p_swapchain_count) {
```
Match the declaration

darksylinc replied:
> Is there a reason to pass these values through the Context creation? You should be able to read the project settings here using GLOBAL_GET (or alternatively, you could fill these values in initialize as well)

No particular reason, I was unsure if I could get the GLOBAL_GET correctly when called afterwards, so I just played safe.

If you tell me it's fine to call `GLOBAL_GET` from the VulkanContext then great, I didn't like passing that plumbing either.

> Since these can only be changed at startup, they should use GLOBAL_DEF_RST_BASIC which will prompt the user to restart if they adjust the setting.
>
> Personally, I think these should be GLOBAL_DEF_RST as they are settings for advanced users only. Removing BASIC means that they setting will only appear if the "advanced" toggle is toggled on.

Ah! Awesome.

clayjohn reviewed (commented):
Great work so far! I left a few comments

AThousandShips commented on main/main.cpp:
```diff
@@ -1993,6 +1995,22 @@ Error Main::setup(const char *execpath, int argc, char *argv[], bool p_second_ph
+		// Alternatively, it makes a lot of sense if set to VSYNC_ADAPTIVE or VSYNC_MAILBOX.
+		// When VSync off, triple buffering increases lag for no (or barely?) improvement and wastes VRAM.
+		// so we force double buffering on that case.
+		if (window_vsync_mode == DisplayServer::VSyncMode::VSYNC_DISABLED)
+			frame_buffering_count = MIN(uint8_t(2u), frame_buffering_count);
```
```suggestion
		if (window_vsync_mode == DisplayServer::VSyncMode::VSYNC_DISABLED) {
			frame_buffering_count = MIN(uint8_t(2u), frame_buffering_count);
		}
```
Always use brackets

AThousandShips commented on platform/windows/vulkan_context_win.cpp:
```diff
@@ -55,7 +55,8 @@ Error VulkanContextWindows::window_create(DisplayServer::WindowID p_window_id, D
 }
 
-VulkanContextWindows::VulkanContextWindows() {
+VulkanContextWindows::VulkanContextWindows(const uint8_t p_frame_count, const uint8_t p_swapchain_count) :
```
```suggestion
VulkanContextWindows::VulkanContextWindows(uint8_t p_frame_count, uint8_t p_swapchain_count) :
```

darksylinc replied:
Is there a reason to remove the constness from function declaration?

I see Godot barely uses const-correctness and it increases the mental load to parse the code by 1000x. I often have to add const to variables while analyzing to help understanding it, and then remove it.

AThousandShips replied:
* The declaration is non-const: `VulkanContextWindows(uint8_t p_frame_count, uint8_t p_swapchain_count);`
* Primitive types don't need constnest, so it isn't relevant, and only make the code more busy

darksylinc replied:
> The declaration is non-const: VulkanContextWindows(uint8_t p_frame_count, uint8_t p_swapchain_count);, at minimum these should match

The reason for that is not to tie the API to a specific detail that may change in the future.

> Primitive types don't need constnest, so it isn't relevant

Yes, it's very relevant. Knowing that `p_swapchain_count` can never change from that line onward is a heavy alleviation on the mental load to parse & understand what a snippet of code is doing.

AThousandShips replied:
> Yes, it's very relevant. Knowing that p_swapchain_count can never change from that line onward is a heavy alleviation on the mental load to parse & understand what a snippet of code is doing.

Generally I agree

darksylinc replied:
> What do you mean? Why shouldn't the header match the source?

The following:

```cpp
// Header
void my_function( int arg );

// Source
void my_function( const int arg )
{
}
```

Indicates:

1. The API/ABI does not impose restrictions.
2. The source uses const to indicate the developers the value won't be changing.
   - If later the that needs to be lifted for some reason, the devs can do it.


The following:

```cpp
// Header
void my_function( const int arg );

// Source
void my_function( const int arg )
{
}
```

1. Now you can't de-const without potentially breaking the API/ABI. You need to use deprecation, e.g. add my_function2.


And finally the following:

```cpp
// Header
void my_function( int arg );

// Source
void my_function( int arg )
{
}
```

1. The API/ABI has no restrictions.
2. Parsing the source is more mentally taxing.

AThousandShips replied:
Sure, it is legal c++, but I don't think we *ever* do this, and it is confusing to have different declaration and definition

Calinou replied:
We use a `p_` prefix for all parameters (and `r_` for out-values), so if you assign a `p_` parameter, it's usually obvious from reading the source (and you probably won't ever be doing this anyway).

clayjohn replied:
I think that is kind of besides the point. ``p_`` means " This is an input parameter _and_ I will not be using this to return a variable. It does not mean "this variable will not be modified in the scope of this function". Adding a const ensures that the parameter won't be changed inside the function. For basic types it may not improve performance as it does with complex types, but that doesn't mean that it isn't beneficial. Readability is often more important than performance for us anyway.

AThousandShips replied:
I think the more important point is:
Do we allow the declaration and definition to differ on the specifics of arguments

I do not believe we have any cases of this anywhere, but I might be wrong

RandomShaper replied:
I'd favor consistency here. The `const` in by-value parameters (primitive type or not) is surprising enough to hamper readability while not letting the reader know anything the `p_` prefix already tells.

I've changed my mind a bit over time on the usage of `const` for locals. If it helps following the code, I'm no longer against it even if I see downsides. However, in cases like this I can't see the benefit.

AThousandShips commented on platform/macos/vulkan_context_macos.mm:
```diff
@@ -55,7 +55,8 @@
 }
 
-VulkanContextMacOS::VulkanContextMacOS() {
+VulkanContextMacOS::VulkanContextMacOS(const uint8_t p_frame_count, const uint8_t p_swapchain_count) :
```
```suggestion
VulkanContextMacOS::VulkanContextMacOS(uint8_t p_frame_count, uint8_t p_swapchain_count) :
```

AThousandShips commented on platform/linuxbsd/x11/vulkan_context_x11.cpp:
```diff
@@ -56,7 +56,8 @@ Error VulkanContextX11::window_create(DisplayServer::WindowID p_window_id, Displ
 }
 
-VulkanContextX11::VulkanContextX11() {
+VulkanContextX11::VulkanContextX11(const uint8_t p_frame_count, const uint8_t p_swapchain_count) :
```
```suggestion
VulkanContextX11::VulkanContextX11(uint8_t p_frame_count, uint8_t p_swapchain_count) :
```

AThousandShips commented on platform/ios/vulkan_context_ios.mm:
```diff
@@ -57,7 +57,9 @@
 }
 
-VulkanContextIOS::VulkanContextIOS() {}
+VulkanContextIOS::VulkanContextIOS(const uint8_t p_frame_count, const uint8_t p_swapchain_count) :
```
```suggestion
VulkanContextIOS::VulkanContextIOS(uint8_t p_frame_count, uint8_t p_swapchain_count) :
```

AThousandShips commented on platform/android/vulkan_context_android.cpp:
```diff
@@ -66,4 +66,8 @@ bool VulkanContextAndroid::_use_validation_layers() {
 	return count > 0;
 }
 
+VulkanContextAndroid::VulkanContextAndroid(const uint8_t p_frame_count, const uint8_t p_swapchain_count) :
```
```suggestion
VulkanContextAndroid::VulkanContextAndroid(uint8_t p_frame_count, uint8_t p_swapchain_count) :
```

darksylinc commented on drivers/vulkan/vulkan_context.h:
```diff
@@ -134,10 +134,15 @@ class VulkanContext {
+	// Initialize to 0 because we don't want it to be used before we initialize and read the config
+	// (this value must stay constant throghout VulkanContext's & RenderingDevice's lifetime)
+	unsigned int frame_count = 0;
+	/// See swapchainImageCount
```
There's no doxygen comments? Ok!

AThousandShips commented on platform/windows/vulkan_context_win.h:
```diff
@@ -44,7 +44,7 @@ class VulkanContextWindows : public VulkanContext {
 	Error window_create(DisplayServer::WindowID p_window_id, DisplayServer::VSyncMode p_vsync_mode, HWND p_window, HINSTANCE p_instance, int p_width, int p_height);
 
-	VulkanContextWindows();
+	VulkanContextWindows(uint8_t p_frame_count, uint8_t p_swapchain_count);
```
This is the declaration I'm talking about

darksylinc commented:
I've squashed and rebased against latest master.

It should be ready for merge now. See if there are any more issues.

Calinou commented on doc/classes/ProjectSettings.xml:
```diff
@@ -826,6 +826,23 @@
+			1 may give you the lowest lag/latency but at the high cost of no parallelism between CPU &amp; GPU.
+			See [url=https://darksylinc.github.io/vsync_simulator/]V-Sync Simulator[/url] to understand the effects of this setting.
+			[b]Note:[/b] This setting is not supported by all APIs.
+			[b]Note:[/b] This property is only read when the project starts.
```
```suggestion
			[b]Note:[/b] This property is only read when the project starts. There is currently no way to change this value at run-time.
```

Calinou commented on doc/classes/ProjectSettings.xml:
```diff
@@ -826,6 +826,23 @@
+			Use Double buffering if V-Sync is off. Triple buffering is a must if you're plan on using V-Sync MAILBOX mode.
+			See [url=https://darksylinc.github.io/vsync_simulator/]V-Sync Simulator[/url] to understand the effects of this setting.
+			[b]Note:[/b] This setting is not supported by all APIs.
+			[b]Note:[/b] This property is only read when the project starts.
```
```suggestion
			[b]Note:[/b] This property is only read when the project starts. There is currently no way to change this value at run-time.
```

Calinou reviewed (approved):
Tested locally on Linux + NVIDIA, it works as expected.

AThousandShips commented on doc/classes/ProjectSettings.xml:
```diff
@@ -826,6 +826,23 @@
+			Sets the number of swapchains (back buffer + front buffer).
+			2 corresponds to double buffering and 3 to triple buffering. A value of 1 is not recommended.
+			Double buffering may give you the lowest lag/latency but if VSync is on and the system can't render at 60 fps, the framerate will go down in multiples of it (e.g. 30 fps, 15, 7.5, etc ). Triple buffering gives you higher framerate (specially if the system can't reach constant 60 fps) at the cost of up to 1 frame of latency (when VSync is On in FIFO mode).
+			Use Double buffering if V-Sync is off. Triple buffering is a must if you're plan on using V-Sync MAILBOX mode.
```
```suggestion
			Use Double buffering if V-Sync is off. Triple buffering is a must if you plan on using V-Sync MAILBOX mode.
```

AThousandShips commented on doc/classes/ProjectSettings.xml:
```diff
@@ -826,6 +826,23 @@
+		</member>
+		<member name="display/window/vsync/swapchain_count" type="int" setter="" getter="" default="3">
+			Sets the number of swapchains (back buffer + front buffer).
+			2 corresponds to double buffering and 3 to triple buffering. A value of 1 is not recommended.
```
```suggestion
			[code]2[/code] corresponds to double buffering and [code]3[/code] to triple buffering. A value of [code]1[/code] is not recommended.
```
Would be good with an explanation for why it is not recommended

AThousandShips commented on doc/classes/ProjectSettings.xml:
```diff
@@ -826,6 +826,23 @@
+		<member name="display/window/vsync/swapchain_count" type="int" setter="" getter="" default="3">
+			Sets the number of swapchains (back buffer + front buffer).
+			2 corresponds to double buffering and 3 to triple buffering. A value of 1 is not recommended.
+			Double buffering may give you the lowest lag/latency but if VSync is on and the system can't render at 60 fps, the framerate will go down in multiples of it (e.g. 30 fps, 15, 7.5, etc ). Triple buffering gives you higher framerate (specially if the system can't reach constant 60 fps) at the cost of up to 1 frame of latency (when VSync is On in FIFO mode).
```
```suggestion
			Double buffering may give you the lowest lag/latency but if VSync is on and the system can't render at 60 fps, the framerate will go down in multiples of it (e.g. 30 fps, 15, 7.5, etc ). Triple buffering gives you higher framerate (specially if the system can't reach a constant 60 fps) at the cost of up to 1 frame of latency (when VSync is On in FIFO mode).
```

AThousandShips replied:
This sentence is not grammatically correct without the article here

AThousandShips commented on drivers/vulkan/vulkan_context.cpp:
```diff
@@ -2766,6 +2768,15 @@ void VulkanContext::set_vsync_mode(DisplayServer::WindowID p_window, DisplayServ
+	// But it may complicate their management.
+	swapchain_desired_count = uint8_t(GLOBAL_DEF_RST("display/window/vsync/swapchain_count", 3u));
+
+	CRASH_COND_MSG(frame_count > MAX_FRAME_LAG, "Frame count out of bounds");
```
```suggestion
	CRASH_COND_MSG(frame_count > MAX_FRAME_LAG, "Frame count out of bounds.");
```

AThousandShips commented on drivers/vulkan/vulkan_context.cpp:
```diff
@@ -2766,6 +2768,15 @@ void VulkanContext::set_vsync_mode(DisplayServer::WindowID p_window, DisplayServ
+	swapchain_desired_count = uint8_t(GLOBAL_DEF_RST("display/window/vsync/swapchain_count", 3u));
+
+	CRASH_COND_MSG(frame_count > MAX_FRAME_LAG, "Frame count out of bounds");
+	CRASH_COND_MSG(frame_count < 1, "Frame count out of bounds");
```
```suggestion
	CRASH_COND_MSG(frame_count < 1, "Frame count out of bounds.");
```

AThousandShips commented on drivers/vulkan/vulkan_context.cpp:
```diff
@@ -2766,6 +2768,15 @@ void VulkanContext::set_vsync_mode(DisplayServer::WindowID p_window, DisplayServ
+
+	CRASH_COND_MSG(frame_count > MAX_FRAME_LAG, "Frame count out of bounds");
+	CRASH_COND_MSG(frame_count < 1, "Frame count out of bounds");
+	CRASH_COND_MSG(swapchain_desired_count < 1, "Frame count out of bounds");
```
```suggestion
	CRASH_COND_MSG(swapchain_desired_count < 1, "Frame count out of bounds.");
```

AThousandShips commented on drivers/vulkan/vulkan_context.cpp:
```diff
@@ -2199,7 +2197,9 @@ Error VulkanContext::_update_swap_chain(Window *window) {
 	};
 
-	for (uint32_t i = 0; i < FRAME_LAG; i++) {
+	CRASH_COND_MSG(frame_count == 0, "Used before initialization");
```
```suggestion
	CRASH_COND_MSG(frame_count == 0, "Used before initialization.");
```

AThousandShips commented on drivers/vulkan/vulkan_context.cpp:
```diff
@@ -1655,7 +1655,10 @@ Error VulkanContext::_create_semaphores() {
 	};
-	for (uint32_t i = 0; i < FRAME_LAG; i++) {
+
+	CRASH_COND_MSG(frame_count == 0, "Used before initialization");
```
```suggestion
	CRASH_COND_MSG(frame_count == 0, "Used before initialization.");
```

darksylinc commented on doc/classes/ProjectSettings.xml:
```diff
@@ -826,6 +826,23 @@
+		<member name="display/window/vsync/buffer_count" type="int" setter="" getter="" default="2">
+			Sets the number of buffers before stalling to wait for the GPU.
+			1 may give you the lowest lag/latency but at the high cost of no parallelism between CPU &amp; GPU.
+			See [url=https://darksylinc.github.io/vsync_simulator/]V-Sync Simulator[/url] to understand the effects of this setting.
```
I believe "understand" is the correct wording here. For starters, there are no examples in that website.

It is an interactive simulator where you can input the desired parameters and see the results, then play with it until how it works "clicks".

AThousandShips replied:
Those are examples, "examples" include things like uses of them, "examples of X", not example code

AThousandShips replied:
Alternatively "for information"

AThousandShips replied:
Perhaps "for an interactive sample" to say what it is

darksylinc replied:
How about

> Try the [url=https://darksylinc.github.io/vsync_simulator/]V-Sync Simulator[/url], an interactive interface that simulates presentation to better understand how it is affected by different variables under various conditions.

?

AThousandShips replied:
Sounds great!

darksylinc replied:
Done!

darksylinc commented:
I had to add more code in light of https://github.com/godotengine/godot/issues/81670 to prevent destroying potentially-uninitialized window' data.

The isolated code changes [can be evaluated here](https://github.com/darksylinc/godot/commit/8ea13496aa1f1017e7ab138b61d7fd9be521fbbf). I wanted to submit them as separate PRs but there were too many merge conflicts otherwise.

darksylinc commented:
I will plea if this PR could be merged and we'll fix any subsequent issue with follow up PRs.

Testing indicates it's working, and it's starting to preclude other fixes.

akien-mga commented:
Clay should be back from holiday by the middle of next week, I expect he'll do a final review pass and approval then.

RandomShaper commented on drivers/vulkan/vulkan_context.cpp:
```diff
@@ -2291,10 +2305,6 @@ Error VulkanContext::prepare_buffers() {
 	VkResult err;
 
-	// Ensure no more than FRAME_LAG renderings are outstanding.
-	vkWaitForFences(device, 1, &fences[frame_index], VK_TRUE, UINT64_MAX);
```
Would the wait still be OK to happen here if it was waiting for the, properly rotated, `frame_index - 1`? Wouldn't that the a more CPU-efficient, given that between `swap_buffers` and the next `prepare_buffers` the application could be using the time for something else than waiting? (Disclaimer: rather than assuming you're wrong in your changes, this is the way for me to understand them.)

darksylinc replied:
If what you're asking is (ignoring performance) if we could wait here doing:

`vkWaitForFences(device, 1, &fences[(frame_index + 1u) % frame_count], VK_TRUE, UINT64_MAX);`

Then the answer is yes, that would be functionally equivalent.

Now if we consider performance, you're asking the comparison:

```cpp
// What you propose
wait( A );
submit( B );

// vs 

// What I wrote
submit( B );
wait( A );
```

In both cases we wait for A. But in my version by the time we're done with the CPU code submission, the GPU may already be done and the wait won't happen.

In fact we should delay `wait( A )` as much as possible (in order to lower the chance we even wait at all). It's just that due to the current complexity of Godot, the current place where I've placed it is the safest spot that I could delay it.

When would we be forced to wait (i.e. cannot delay any longer)?

 - We need to destroy a vulkan handle (e.g. a framebuffer, render pass, possibly a shader & pso if they were phased out N frames ago).
 - We need to map a buffer for reading or writing.
 - We need to reuse (reset) a command buffer for preparing commands.

RandomShaper replied:
So, you're doing this:
```
swap_buffers():
	submit(frame_index)
	frame_index += (frame_index + 1) % N
	wait(frame_index)
```

Can't that be changed to this?:
```
prepare_buffers():
    wait(frame_index)

swap_buffers():
	submit(frame_index)
	frame_index += (frame_index + 1) % N
```

I think stuff happens between `swap_buffers()` and the next `prepare_buffers()`. Wouldn't this allow that stuff to use the CPU in cases where it wouldn't have any choice but wait for the GPU? In other words, I think this is maybe the chance you mentioned about waiting as late as possible.

RandomShaper replied:
Makes sense. I was about to suggest moving the call to `_free_pending_resources()` to a safe spot, but if there are other potential usafe operations between, I'm happy with the change. I hope we can polish that part down to perfection in the future.

thygrrr replied:
Hmm, I don't think that's what's happening here.

YES, the original order is:
``` 
	vkWaitForFences(device, 1, &fences[frame_index], VK_TRUE, UINT64_MAX);
	vkResetFences(device, 1, &fences[frame_index]);
```
but this is in prepare_buffers, which happens much later. The vkResetFences should probably go where it does in your PR (much cleaner), but the wait should happen right at the beginning of prepare_buffres.

`frame_index` is incremented after swap_buffers, meaning that each time we arrive here, we have already done all other work and are just getting ready to `prepare_screen_for_drawing`.

We're basically waiting at the last possible moment to acquire the image, just before rendering.

(technically, there is an even later time, this is before the final blit, but that's terrible for tiled renderers and mobile, and might degrade performance on desktop for slightly less micro stutter)


I'm currently examining an issue that frames actually get shown out of order, and even repeatedly. I don't know the cause yet, but I suspect keeping separate "buffers" around instead of allocating one set per swapchain image is likely part of the problem. (not a problem of this PR, but maybe you have some insights on that?)

darksylinc replied:
> but the wait should happen right at the beginning of prepare_buffers

That is the ideal, yes. But as I explained in [my comment](https://github.com/godotengine/godot/pull/80566#discussion_r1328808985) if you do that right now, you'll start deleting resources that must not be deleted without waiting first.

And even if we fixed that, Godot has no mechanism in place to prevent or detect nobody in future PRs accidentally introduces dangerous operations before the wait (that must happen after the wait).

RandomShaper commented on doc/classes/ProjectSettings.xml:
```diff
@@ -827,6 +827,23 @@
+			[b]Note:[/b] This property is only read when the project starts. There is currently no way to change this value at run-time.
+		</member>
+		<member name="display/window/vsync/swapchain_count" type="int" setter="" getter="" default="3">
+			Sets the number of swapchains (back buffer + front buffer).
```
I'm not sure (maybe because of ignorance) about this wording. Isn't there a single swapchin consisting of 1 front buffer plus N back buffers?

darksylinc replied:
This is always confusing because D3D11 asks for backbuffer count (backbuffer = 2 means triple buffer because you have to add the implicit front buffer) while Vulkan asks for swapchain count (where swapchaincount = 2 means double buffer because you're including the front one).

RandomShaper replied:
What do you think about naming this `swapchain_image_count`?

TempoLabGames replied:
I may be fundamentally misunderstanding something, but is it ever actually useful for this to be configurable? The description seems to suggest this is completely determined by other settings. For `VSYNC_DISABLED` the only sensible value is 2. For `VSYNC_ENABLED` and `VSYNC_ADAPTIVE` this feels like it ought to be `buffer_count + 1`. For `VSYNC_MAILBOX` the only sensible value is 3.

TempoLabGames replied:
Ah, I was definitely misunderstanding. Your tool is really useful, and the info in the README cleared everything up for me. The documentation you've added in this PR makes perfect sense. Thank you very much, and sorry for derailing!

darksylinc commented on platform/android/vulkan_context_android.h:
```diff
@@ -43,7 +43,7 @@ class VulkanContextAndroid : public VulkanContext {
 	Error window_create(ANativeWindow *p_window, DisplayServer::VSyncMode p_vsync_mode, int p_width, int p_height);
 
-	VulkanContextAndroid() = default;
+	VulkanContextAndroid();
```
It's a vestige of an older version of this PR that required to modify VulkanContextAndroid constructor.

darksylinc commented on drivers/vulkan/vulkan_context.h:
```diff
@@ -153,7 +158,7 @@ class VulkanContext {
 		SwapchainImageResources *swapchain_image_resources = VK_NULL_HANDLE;
 		VkPresentModeKHR presentMode = VK_PRESENT_MODE_FIFO_KHR;
-		VkSemaphore image_acquired_semaphores[FRAME_LAG];
+		VkSemaphore image_acquired_semaphores[MAX_FRAME_LAG];
```
Does that actually make it zero-initialized? I always get bitten back by C++11 rules of {}.

RandomShaper replied:
Yes, it does. Syntax-wise looks like saying nothing indeed. 😀

darksylinc replied:
What I meant is that the C++11 is bananas because {} does **_not always_** zero-initialize, which can be ridiculously misleading.

darksylinc commented on drivers/vulkan/rendering_device_vulkan.h:
```diff
@@ -1038,13 +1038,11 @@ class RenderingDeviceVulkan : public RenderingDevice {
 	bool local_device_processing = false;
 
-	void _free_pending_resources(int p_frame);
+	void _free_pending_resources(unsigned int p_frame);
```
I agree.

I was just following the same trend of keeping the type frame_index was using.

DarioSamo commented on drivers/vulkan/vulkan_context.cpp:
```diff
@@ -2767,6 +2783,15 @@ void VulkanContext::set_vsync_mode(DisplayServer::WindowID p_window, DisplayServ
+
+	CRASH_COND_MSG(frame_count > MAX_FRAME_LAG, "Frame count out of bounds.");
+	CRASH_COND_MSG(frame_count < 1, "Frame count out of bounds.");
+	CRASH_COND_MSG(swapchain_desired_count < 1, "Frame count out of bounds.");
```
You probably want to make the error message here match the condition to describe that swap chain desired count is the problem.

AllianceDarkSylinc replied:
Reworded as suggested. I agree with all of it. It's much more clear to the user.

The old wording was from the initial version of the code which was not taking the input directly from `GLOBAL_DEF_RST`.

darksylinc replied:
Reworded as suggested. I agree with all of it. It's much more clear to the user.

The old wording was from the initial version of the code which was not taking the input directly from `GLOBAL_DEF_RST`.

TempoLabGames commented:
I've built and tested this using https://github.com/TempoLabGames/UnityRenderingLatencyDemo

Summary: This change makes a significant difference to ExclusiveFullscreen, but no difference to Windowed/Fullscreen.

I've included measurements against Godot 4.1.1, your latest commit (my changes to build the .NET editor are available at https://github.com/TempoLabGames/godot/commits/matias-vulkan-sync-fix), and Unity's Vulkan and Direct3D11 implementations for reference.

(Note Godot's ExclusiveFullscreen is equivalent to Unity's FullScreenWindow; **not** Unity's ExclusiveFullScreen.)

Godot 4.1.1 Forward+:
* Windowed: 2-3 frames
* ExclusiveFullscreen: 5 frames

Godot 4302a80 Forward+ (buffer_count=2, swapchain_count=3):
* Windowed: 2-3 frames
* ExclusiveFullscreen: 4 frames

Godot 4302a80 Forward+ (buffer_count=2, swapchain_count=2):
* Windowed: 2-3 frames
* ExclusiveFullscreen: 3 frames

Godot 4302a80 Forward+ (buffer_count=1, swapchain_count=2):
* Windowed: 2-3 frames
* ExclusiveFullscreen: 2 frames (or slightly less)

Unity 2021.3 Vulkan (QualitySettings.maxQueuedFrames=1):
* Windowed: 3 frames
* FullScreenWindow: 5 frames

Unity 2021.3 Direct3D11 (QualitySettings.maxQueuedFrames=1):
* Windowed: 2 frames
* FullScreenWindow: 1 frame

darksylinc committed:
Fix Vulkan CPU/GPU synchronization

The original code had two bugs:

1. It assumed Swapchain count can be used for proper CPU/GPU
synchronization (it's not)
2. It used a double buffer scheme (FRAME_LAG = 2) but the
vkWaitForFences/vkResetFences calls were in the wrong order and in the
wrong place
3. The swapchain bug forced a quad-buffer scheme with roughly 1 or 2
frames of wait; which hid all the issues

This commit gets rid of RenderingDeviceVulkan::frame_count &
RenderingDeviceVulkan::frame which should've never existed and makes
everything depend on VulkanContext::frame_index &
VulkanContext::FRAME_LAG

Add ability to tweak FRAME_LAG from within Project Settings
Adds new setting display/window/vsync/buffer_count
Adds new setting display/window/vsync/swapchain_count

See https://github.com/godotengine/godot/issues/80550 for details

------------------------------------------------------------------

Fix comment grammar
Use Godot's functions instead of std

------------------------------------------------------------------

Additionally handle inconsistent state when destroying windows:
window.image_acquired_semaphores[i] contains uninitialized values.

This means that if VulkanContext::_update_swap_chain had failed early
(this can happen for valid reasons and due to unexpected failures) then
_clean_up_swap_chain will try to destroy an invalid handle.
This fix always sets the handle to VK_NULL_HANDLE.

Likewise perform the checks in various other resources in
_clean_up_swap_chain in case the resources were partially initialized
and are now in an inconsistent state. The API should guarantee the call
to be valid if the handle was VK_NULL_HANDLE, but I don't trust driver
implementations to handle that case.

Affects #81670

Co-authored-by: Clay John <claynjohn@gmail.com>

darksylinc commented:
I've rebased this branch with latest master and incorporated the latest suggestions.

TisFoolish commented:
> (Note Godot's ExclusiveFullscreen is equivalent to Unity's FullScreenWindow; **not** Unity's ExclusiveFullScreen.)

@TempoLabGames Is Unity's ExclusiveFullScreen at 0 frames of delay?

I assume these tests are on Windows (which version?) because of the inclusion of the D3D11 test, unless you're running this through proton? 

What GPU are you using?

TempoLabGames commented:
> Is Unity's ExclusiveFullScreen at 0 frames of delay?

@TisFoolish No, the test is relative to the mouse cursor, so the theoretical best latency you can achieve with VSync on (assuming you're allowing a full frame to generate the frame) is one frame. There is generally no need to use Unity's ExclusiveFullScreen any more since modern versions of Windows allow borderless full screen windows to bypass DWM in the same way an exclusive full screen application would. See https://learn.microsoft.com/en-us/windows/win32/direct3ddxgi/for-best-performance--use-dxgi-flip-model for more details.

> I assume these tests are on Windows (which version?) because of the inclusion of the D3D11 test, unless you're running this through proton?
>
> What GPU are you using?

Sorry, I should have mentioned. I'm using Windows 10 with an NVIDIA GeForce GTX 660.

AThousandShips commented on platform/ios/vulkan_context_ios.mm:
```diff
@@ -57,7 +57,8 @@
 }
 
-VulkanContextIOS::VulkanContextIOS() {}
+VulkanContextIOS::VulkanContextIOS() {
+}
```
```suggestion
VulkanContextIOS::VulkanContextIOS() {}
```
Restore this leftover

clayjohn commented on drivers/vulkan/rendering_device_vulkan.cpp:
```diff
@@ -9015,10 +9049,11 @@ void RenderingDeviceVulkan::initialize(VulkanContext *p_context, bool p_local_de
 		vmaCreateAllocator(&allocatorInfo, &allocator);
 	}
 
+	const uint32_t frame_count = context->get_frame_count();
```
When using a local rendering device we will end up allocating resources for many frames even though we only need one.

darksylinc replied:
You're thinking this in terms of "visual frames" but frame_count what really does is tracking synchronization boundaries.

If we only have one internal "frame", then after every submit we must stall.

Think of CPU and GPU as two horses racing in circles, and the GPU horse must ALWAYS stay behind the CPU horse.
And the CPU horse must NEVER overtake the GPU horse by more than N laps (N = frame_count).
But the GPU horse can only start galloping when the CPU horse raider says "submit".

Even if the user never ever sees a single window, we need a way to prevent the CPU horse from overtaking the GPU horse. If frame_count = 1, that's basically the same as the CPU horse saying "submit" and then hard-breaking until the GPU horse catches up.

clayjohn commented on drivers/vulkan/rendering_device_vulkan.cpp:
```diff
@@ -8765,6 +8800,7 @@ void RenderingDeviceVulkan::submit() {
 
 	_finalize_command_bufers();
 
+	const uint32_t frame = context->get_frame_index();
 	VkCommandBuffer command_buffers[2] = { frames[frame].setup_command_buffer, frames[frame].draw_command_buffer };
```
The design of the LocalRenderingDevice is essentially exposing a RenderingDevice without a swapchain or concept of a frame. Instead of swapping buffers, users are expected to call ``submit()`` and ``sync()`` and convenient times so that they can schedule work as they want to. 

I think having a global ``frame`` index can make sense for multiple windows and such, but it doesn't really make sense for the local rendering device.

darksylinc replied:
> Consider the following example:

No. That must never happen.

If the frame index is about to be increased, we **must** first submit and a VkFence issued so we can wait on it.

At least when it comes to what VulkanContext considers a 'frame'. If the higher level interface wants to render two frames and submit them to VulkanContext as a single 'frame' that's alright.

clayjohn replied:
> > Consider the following example:
> 
> No. That must never happen.
> 
> If the frame index is about to be increased, we **must** first submit and a VkFence issued so we can wait on it.
> 
> At least when it comes to what VulkanContext considers a 'frame'. If the higher level interface wants to render two frames and submit them to VulkanContext as a single 'frame' that's alright.

I think you are misunderstanding what a local rendering device is. The entire point of it is to have a place to submit commands over multiple frames. For work that is okay to be run in the background. When the user is done recording, they call submit, then they cannot record or submit until they call sync. 

Logically, this doesn't map onto the concept of frames. You have a single queue, a single command_buffer. You submit when you are done recording, and you sync before you start recording again. There is no need for multiple frames, and indeed, having multiple frames makes things way more complicated as you need to start to manage what work was done in which frame. 

The idea behind the LocalRenderingDevice is that when users need per-frame stuff, they just use the global rendering device, if they are okay having something run in the background over multiple frames, they can use the localrenderingdevice

RandomShaper replied:
If it helps, 'local rendering device' is a headless thing.

RandomShaper commented on doc/classes/ProjectSettings.xml:
```diff
@@ -830,6 +830,23 @@
 		<member name="display/window/subwindows/embed_subwindows" type="bool" setter="" getter="" default="true">
 			If [code]true[/code] subwindows are embedded in the main window.
 		</member>
+		<member name="display/window/vsync/buffer_count" type="int" setter="" getter="" default="2">
+			Sets the number of buffers before stalling to wait for the GPU.
+			1 may give you the lowest lag/latency but at the high cost of no parallelism between CPU &amp; GPU.
+			Try the [url=https://darksylinc.github.io/vsync_simulator/]V-Sync Simulator[/url], an interactive interface that simulates presentation to better understand how it is affected by different variables under various conditions.
+			[b]Note:[/b] This setting is not supported by all APIs.
+			[b]Note:[/b] This property is only read when the project starts. There is currently no way to change this value at run-time.
+		</member>
+		<member name="display/window/vsync/swapchain_count" type="int" setter="" getter="" default="3">
+			Sets the number of swapchains (back buffer + front buffer).
+			[code]2[/code] corresponds to double buffering and [code]3[/code] to triple buffering. A value of [code]1[/code] is not recommended.
+			Double buffering may give you the lowest lag/latency but if VSync is on and the system can't render at 60 fps, the framerate will go down in multiples of it (e.g. 30 fps, 15, 7.5, etc ). Triple buffering gives you higher framerate (specially if the system can't reach a constant 60 fps) at the cost of up to 1 frame of latency (when VSync is On in FIFO mode).
+			Use Double buffering if V-Sync is off. Triple buffering is a must if you plan on using V-Sync MAILBOX mode.
+			Try the [url=https://darksylinc.github.io/vsync_simulator/]V-Sync Simulator[/url], an interactive interface that simulates presentation to better understand how it is affected by different variables under various conditions.
+			[b]Note:[/b] This setting is not supported by all APIs.
+			[b]Note:[/b] This property is only read when the project starts. There is currently no way to change this value at run-time.
+			[b]Note:[/b] Some platforms may restrict the actual value.
+		</member>
```
I'm suggesting moving these settings to `rendering/rendering_device/vsync`, as only `RenderingDevice` backends offer such a degree of control. I'm also suggesting alternative names and descriptions, for clarity (e.g., 'buffers' alone is a term already used to name another thing in the rendering API).

```suggestion
		<member name="rendering/rendering_device/vsync/frame_queue_size" type="int" setter="" getter="" default="2">
			The number of frames to track on the CPU side before stalling to wait for the GPU.
			1 may give you the lowest lag/latency but at the high cost of no parallelism between CPU &amp; GPU.
			Try the [url=https://darksylinc.github.io/vsync_simulator/]V-Sync Simulator[/url], an interactive interface that simulates presentation to better understand how it is affected by different variables under various conditions.
			[b]Note:[/b] This setting is not supported by all APIs.
			[b]Note:[/b] This property is only read when the project starts. There is currently no way to change this value at run-time.
		</member>
		<member name="rendering/rendering_device/vsync/swapchain_image_count" type="int" setter="" getter="" default="3">
			The number of images the swapchain will consist of (back buffers + front buffer).
			[code]2[/code] corresponds to double-buffering and [code]3[/code] to triple-buffering. A value of [code]1[/code] is not recommended.
			Double-buffering may give you the lowest lag/latency but if V-Sync is on and the system can't render at 60 fps, the framerate will go down in multiples of it (e.g. 30 fps, 15, 7.5, etc ). Triple buffering gives you higher framerate (specially if the system can't reach a constant 60 fps) at the cost of up to 1 frame of latency, with [member DisplayServer.VSYNC_ENABLED] (FIFO).
			Use double-buffering with [member DisplayServer.VSYNC_ENABLED]. Triple-buffering is a must if you plan on using [member DisplayServer.VSYNC_MAILBOX] mode.
			Try the [url=https://darksylinc.github.io/vsync_simulator/]V-Sync Simulator[/url], an interactive interface that simulates presentation to better understand how it is affected by different variables under various conditions.
			[b]Note:[/b] This setting is not supported by all APIs.
			[b]Note:[/b] This property is only read when the project starts. There is currently no way to change this value at run-time.
			[b]Note:[/b] Some platforms may restrict the actual value.
		</member>
```

NotRabug50 commented:
https://github.com/godotengine/godot/assets/78092286/a3edc681-cdb6-4ce8-8c39-b78d92e31f9f

i think you guys still need to test before merging. This bug still can happen if you overload too much

thygrrr commented on drivers/vulkan/vulkan_context.h:
```diff
@@ -154,7 +159,7 @@ class VulkanContext {
 		SwapchainImageResources *swapchain_image_resources = VK_NULL_HANDLE;
 		VkPresentModeKHR presentMode = VK_PRESENT_MODE_FIFO_KHR;
-		VkSemaphore image_acquired_semaphores[FRAME_LAG];
+		VkSemaphore image_acquired_semaphores[MAX_FRAME_LAG] = {};
```
Could you elaborate why multiple acquired semaphores are needed per window? 
AFAIK, only one swapchain image is generally acquired at a time. (this is actually an issue I was looking into for https://github.com/godotengine/godot/issues/84137 and in fact, I saw (in superficial testing) no detriment in keeping just 1 image_aquired_semaphore per Window.

(I'm quite possibly bad and wrong about this)

darksylinc replied:
Hi!

I had the same question actually!

Here's the thing:

1. Indeed just one semaphore for all windows is how I think should be properly done (unless you're careful with barriers and want to explicitly start doing Compute Work using async compute for the next frame while the current frame is still being worked on)
2. In OgreNext we use one semaphore and funny enough, that triggered driver bugs for various drivers (which have been long been fixed)
3. Official and unofficial samples use one semaphore per swapchain (this is most likely why the drivers didn't catch that bug for OgreNext...), and since I saw that Godot was doing the same, I decided not to change this.

But yes, for the record for desktop\*, I am more comfortable with one semaphore for everything because that's how I think proper synchronization should be.

\* I say desktop because I'm having timing issues on Android that I can't fix in OgreNext and I never tested the theory of whether they're caused because of having one semaphore for everything. But on the other hand, Android has a lot of problems of all sorts.

thygrrr commented on drivers/vulkan/vulkan_context.cpp:
```diff
@@ -1680,7 +1680,10 @@ Error VulkanContext::_create_semaphores() {
 	};
-	for (uint32_t i = 0; i < FRAME_LAG; i++) {
+
+	CRASH_COND_MSG(frame_count == 0, "Used before initialization.");
```
Wonderful! I ran into that same issue trying a provisional fix to what your PR does, this seems much cleaner than the current logic on master.

thygrrr commented on doc/classes/ProjectSettings.xml:
```diff
@@ -830,6 +830,23 @@
+			[code]2[/code] corresponds to double buffering and [code]3[/code] to triple buffering. A value of [code]1[/code] is not recommended.
+			Double buffering may give you the lowest lag/latency but if VSync is on and the system can't render at 60 fps, the framerate will go down in multiples of it (e.g. 30 fps, 15, 7.5, etc ). Triple buffering gives you higher framerate (specially if the system can't reach a constant 60 fps) at the cost of up to 1 frame of latency (when VSync is On in FIFO mode).
+			Use Double buffering if V-Sync is off. Triple buffering is a must if you plan on using V-Sync MAILBOX mode.
+			Try the [url=https://darksylinc.github.io/vsync_simulator/]V-Sync Simulator[/url], an interactive interface that simulates presentation to better understand how it is affected by different variables under various conditions.
```
I don't think this link should go into the Godot documentation.

Calinou replied:
> I don't think this link should go into the Godot documentation.

Why not?

AThousandShips replied:
```suggestion
			Try the [url=https://darksylinc.github.io/vsync_simulator/]V-Sync Simulator[/url], an interactive interface that simulates presentation, to better understand how it is affected by different variables under various conditions.
```
Add punctuation to separate clause

thygrrr commented on drivers/vulkan/vulkan_context.cpp:
```diff
@@ -1960,18 +1977,13 @@ Error VulkanContext::_update_swap_chain(Window *window) {
-	if (desiredNumOfSwapchainImages < surfCapabilities.minImageCount) {
-		desiredNumOfSwapchainImages = surfCapabilities.minImageCount;
-	}
+	uint32_t desiredNumOfSwapchainImages = MAX(surfCapabilities.minImageCount, swapchain_desired_count);
```
Some sources I found seem to state it's sensible to also lower bound this by `surfCapabilities.minImageCount+1` (this might, on particularly conservative implementations, relax the contention for driver resources a bit when acquiring the next image; the common case here is still that the default 3 images are the value that you are gonna get)

darksylinc replied:
I found all references to this info trace back to the [Vulkan Tutorial](https://vulkan-tutorial.com/Drawing_a_triangle/Presentation/Swap_chain).

After research I found the following:

1. The OSes that have sync issues (like X11 on Linux) already set minImageCount to a large enough value like 4 or 5.
2. I suspect it is highly likely that the tutorial confused "wait on the driver to complete internal operations" with how VSync is supposed to work. This is known and can be observed in [my VSync Simulator](https://darksylinc.github.io/vsync_simulator/). If you set swapchain_count to 2 and the rendering time is higher than the refresh rate, then the entire framerate will alternate to an integer divisor of the refresh rate (e.g. 60fps, 30 fps, 15 fps). Depending on how the OS compositor internally works, this might happen with swapchain_count > 2.
    - Seeing the [other ticket](https://github.com/godotengine/godot/issues/71795), it is clear some users explicitly want this behavior (because if the CPU & GPU are fast enough to not miss any VBLANK, then latency is minimum).

thygrrr commented on platform/ios/vulkan_context_ios.mm:
```diff
@@ -57,7 +57,8 @@
 }
 
-VulkanContextIOS::VulkanContextIOS() {}
+VulkanContextIOS::VulkanContextIOS() {
```
unnecessary addition of whitespace.

AThousandShips commented on doc/classes/ProjectSettings.xml:
```diff
@@ -830,6 +830,23 @@
+		<member name="display/window/vsync/buffer_count" type="int" setter="" getter="" default="2">
+			Sets the number of buffers before stalling to wait for the GPU.
+			1 may give you the lowest lag/latency but at the high cost of no parallelism between CPU &amp; GPU.
+			Try the [url=https://darksylinc.github.io/vsync_simulator/]V-Sync Simulator[/url], an interactive interface that simulates presentation to better understand how it is affected by different variables under various conditions.
```
```suggestion
			Try the [url=https://darksylinc.github.io/vsync_simulator/]V-Sync Simulator[/url], an interactive interface that simulates presentation, to better understand how it is affected by different variables under various conditions.
```

Opalie commented:
Hi, would it be possible to get the build artifacts of this PR ? They are sadly not available anymore. Thank you very much

vpellen commented:
Is this just waiting on review, or has it effectively been superseded by #87340?

DarioSamo commented:
> Is this just waiting on review, or has it effectively been superseded by #87340?

It'd effectively be superseded by that, so if you're waiting on this one I'd suggest confirming if that PR fixes the problem as well. The PRs will be completely incompatible and Matias has reviewed that I implemented what he detected correctly in that PR as well, as well as fetch the settings/doc changes from it.

vpellen commented:
Just a minor follow up, I did test the other PR and it doesn't seem to fix the obnoxious stuttering, although the curious thing is that neither of the directx implementations (both the angle compatibility mode and the recent dx12 stuff) suffer from the stuttering issues, so this may be some insufferable nvidia vulkan/opengl issue. The hunt for "why does vsync cause so much misery" continues. Still, at least there are options now.

AThousandShips commented on doc/classes/ProjectSettings.xml:
```diff
@@ -830,6 +830,23 @@
+		<member name="display/window/vsync/swapchain_count" type="int" setter="" getter="" default="3">
+			Sets the number of swapchains (back buffer + front buffer).
+			[code]2[/code] corresponds to double buffering and [code]3[/code] to triple buffering. A value of [code]1[/code] is not recommended.
+			Double buffering may give you the lowest lag/latency but if VSync is on and the system can't render at 60 fps, the framerate will go down in multiples of it (e.g. 30 fps, 15, 7.5, etc ). Triple buffering gives you higher framerate (specially if the system can't reach a constant 60 fps) at the cost of up to 1 frame of latency (when VSync is On in FIFO mode).
```
```suggestion
			Double buffering may give you the lowest lag/latency but if VSync is on and the system can't render at 60 fps, the framerate will go down in multiples of it (e.g. 30 fps, 15, 7.5, etc.). Triple buffering gives you higher framerate (specially if the system can't reach a constant 60 fps) at the cost of up to 1 frame of latency (when VSync is On in FIFO mode).
```

akien-mga commented:
Superseded by https://github.com/godotengine/godot/pull/87340. Thanks for the great work on this!
